Update the spec according to the feedback (#78)
riscv-non-isa/riscv-external-debug-security: chapter2.adoc

[[mdbgctl]]
==== M-mode Debug Control

A state element in each hart, named `mdbgen`, is introduced to control the debuggability of M-mode for each hart as depicted in <<extdbg>>. When `mdbgen` is set to 1, the following rules apply:

- The <<dbgaccpriv, debug access privilege>> for the hart is M-mode. Abstract Commands, including "Quick Access", and Program Buffer execution operate with M-mode privilege
- The <<dbops, debug operations>> are permitted when the hart executes in any privilege mode.

When `mdbgen` is set to 0, external debug is disallowed in M-mode. See <<sdsecextdbg>> for how this impacts hart behavior in M-mode. 

[NOTE]
The `mdbgen` may be controlled through various methods, such as a new input port to the hart, a handshake with the system Root of Trust (RoT), or other methods. The implementation can choose to group several harts together and use one signal to drive their `mdbgen` state or assign each hart its own dedicated state. For example, a homogeneous computing system can use a signal to drive all `mdbgen` state to enforce a unified debug policy across all harts.

[NOTE]

> The `mdbgen` of the non-RoT hart shall be managed by the RoT firmware. The `mdbgen` of the RoT hart (if present) shall be managed by SOC hardware. It can be OTP/life-cycle fuse optionally with authentication module.

Author: When and how to set `mdbgen` can vary across implementations. Some implementations might design a specific handshake protocol to enable `mdbgen`, while others might manage it directly through lifecycle. It might be more appropriate to leave the specifics of enabling `mdbgen` to the implementation.

> The `mdbgen` of the non-RoT hart shall be managed by the RoT firmware. The `mdbgen` of the RoT hart (if present) shall be managed by SOC hardware. It can be OTP/life-cycle fuse optionally with authentication module.

Reviewer: > Should there be guidance on when the RoT should set the mdbgen bit for a given hart?

IMO, we should, but probably in a different document

[[submdbgctl]]
==== Supervisor Domain Debug Control

> The Smsdedbg extension cite:[smmtt] introduces `sdedbgalw` field (bit 7) in CSR <<Sdseccsr,msdcfg>> to control the debuggability of supervisor domains. When `mdbgen`=0, the `sdedbgalw` field determines both the debug allowed privilege modes and the <<dbgaccpriv, debug access privilege>>, as illustrated in <<dbgpriv>>. When disallowed, see <<sdsecextdbg>> for the impact on hart behavior in debug disallowed modes. 

Reviewer 2: I don't understand the new sentence.  When what is disallowed?  Did you mean to say "When external debug is not allowed in the current privilege mode"?  I'm not sure this sentence is needed, and may add confusion.

Author: Removed.

[[dbgdisallowed]]
* The hart will not enter Debug Mode. Halt requests will remain pending until debug is allowed.  
* Triggers with `action`=1 (enter Debug Mode) will not match or fire.

The subsequent subsections describe how external debug is authorized by <<mdbgctl, M-mode debug control>> and <<submdbgctl, supervisor domain debug control>>.

[[mdbgctl]]
==== M-mode Debug Control

A state element in each hart, named `mdbgen`, is introduced to control the debuggability of M-mode for each hart as depicted in <<extdbg>>. When `mdbgen` is set to 1, the following rules apply:

- The <<dbgaccpriv, debug access privilege>> for the hart is M-mode. Abstract Commands, including "Quick Access", and Program Buffer execution operate with M-mode privilege
- The <<dbops, debug operations>> are permitted when the hart executes in any privilege mode.

When `mdbgen` is set to 0, external debug is disallowed in M-mode. See <<sdsecextdbg>> for how this impacts hart behavior in M-mode. 

[NOTE]

> The `mdbgen` may be controlled through various methods, such as a new input port to the hart, a handshake with the system Root of Trust (RoT), or other methods. The implementation can choose to group several harts together and use one signal to drive their `mdbgen` state or assign each hart its own dedicated state. For example, a homogeneous computing system can use a signal to drive all `mdbgen` state to enforce a unified debug policy across all harts.

Reviewer: ```suggestion
The management of `mdgben` depends on its implementation, which is beyond the scope of this specification. However, it is recommended that implementations follow security principles when implementing `mdgben` for harts in SOC. The `mdbgen` of a non-RoT hart should be controlled by RoT, while the `mdbgen` of a RoT hart should be controlled by hardware or immutable code, such as ROM.  
```

[[dbgdisallowed]]
* The hart will not enter Debug Mode. Halt requests will remain pending until debug is allowed.  
* Triggers with `action`=1 (enter Debug Mode) will not match or fire.

The subsequent subsections describe how external debug is authorized by <<mdbgctl, M-mode debug control>> and <<submdbgctl, supervisor domain debug control>>.

[NOTE]

> A pending request to enter Debug Mode can dynamically change from a disallowed state to an allowed state due to updates in debug controls. For example, once the software completes executing confidential code, it can grant debuggability for an external debugger, and the hart will acknowledge the pending halt request at that point. 

Reviewer: May need more clarification on "dynamically" change, or just remove. Do we want to say debug control may change anytime during execution?

Reviewer: Probably just remove..

Author: sure, removed the note in last commit

[[mdbgctl]]
==== M-mode Debug Control

A state element in each hart, named `mdbgen`, is introduced to control the debuggability of M-mode for each hart as depicted in <<extdbg>>. When `mdbgen` is set to 1, the following rules apply:

- The <<dbgaccpriv, debug access privilege>> for the hart is M-mode. Abstract Commands, including "Quick Access", and Program Buffer execution operate with M-mode privilege
- The <<dbops, debug operations>> are permitted when the hart executes in any privilege mode.

When `mdbgen` is set to 0, external debug is disallowed in M-mode. See <<sdsecextdbg>> for how this impacts hart behavior in M-mode. 

[NOTE]
The `mdbgen` may be controlled through various methods, such as a new input port to the hart, a handshake with the system Root of Trust (RoT), or other methods. The implementation can choose to group several harts together and use one signal to drive their `mdbgen` state or assign each hart its own dedicated state. For example, a homogeneous computing system can use a signal to drive all `mdbgen` state to enforce a unified debug policy across all harts.

[NOTE]

> For M-mode firmware debugging with protected ROM code, assert `mdbgen` only after the ROM code is completed. For example, the `mdbgen` of the non-RoT hart can be controlled by the RoT. The `mdbgen` of the RoT hart, if present, can be managed by external hardware like OTP/life-cycle fuse, with optional authentication. Establishing a chain of trust (CoT) allows M-mode firmware debugging without exposing ROM code to an unauthorized debugger.

Reviewer: ```suggestion
The `mdbgen` of the non-RoT hart shall be managed by the RoT firmware. The `mdbgen` of the RoT hart (if present) shall be managed by SOC hardware. It can be OTP/life-cycle fuse optionally with authentication module.
```

[[submdbgctl]]
==== Supervisor Domain Debug Control

> The Smsdedbg extension cite:[smmtt] introduces `sdedbgalw` field (bit 7) in CSR <<Sdseccsr,msdcfg>>, to control the debuggability of supervisor domains. When `mdbgen`=0, the `sdedbgalw` field determines both the debug allowed privilege modes and the <<dbgaccpriv, debug access privilege>>, as illustrated in <<dbgpriv>>. When disallowed, see <<sdsecextdbg>> for the impact on hart behavior in debug disallowed modes. 

Reviewer: ```suggestion
The Smsdedbg extension cite:[smmtt] introduces `sdedbgalw` field (bit 7) in CSR <<Sdseccsr,msdcfg>> to control the debuggability of supervisor domains. When `mdbgen`=0, the `sdedbgalw` field determines both the debug allowed privilege modes and the <<dbgaccpriv, debug access privilege>>, as illustrated in <<dbgpriv>>. When disallowed, see <<sdsecextdbg>> for the impact on hart behavior in debug disallowed modes. 
```

The external trigger outputs (with `action` = 8/9) will not fire or match when the privilege level of the hart exceeds debug allowed privilege as specified in <<dbgpriv>>.

> The external trigger inputs (tmexttrigger) can be driven by any input signals, e.g. the external trigger output from another hart, interrupt signals, etc. The initiators of these signals are responsible for determining whether the signal is allowed to assert. The hart will not acknoledge the input until it is in a debug allowed state. For example, if the external trigger input of hart _i_ is connected to external trigger output of hart _j_, the assertion of output signal from hart _j_ is determined by its own allowed privilege level for debug. The hart _i_ will response according to `tmexttrigger.action` only when it is in a debug allowed state and hart _j_ asserts the output signal.

Reviewer 3: ```suggestion
The external trigger inputs (tmexttrigger) can be driven by any input signals, e.g. the external trigger output from another hart, interrupt signals, etc. The initiators of these signals are responsible for determining whether the signal is allowed to assert. The hart will not acknowledge the input until it is in a debug allowed state. For example, if the external trigger input of hart _i_ is connected to external trigger output of hart _j_, the assertion of output signal from hart _j_ is determined by its own allowed privilege level for debug. Hart _i_ will respond according to `tmexttrigger.action` only when it is in a debug allowed state and hart _j_ asserts the output signal.
```
Or should the last sentence say "If `tmexttrigger.action` is 1, hart _i_ will halt only when it is in a debug allowed state and hart _j_ asserts the output signal."?  Because, for instance, action=0 shouldn't be affected by any of this.

Author: Agree. Updated the last sentence in last commit
